framework/src/media/demux/mpeg2ts/ParseManager.cpp

```diff
@@ -0,0 +1,286 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************/
+
+#include <debug.h>
+#include "Mpeg2TsTypes.h"
+#include "Section.h"
+#include "SectionParser.h"
+#include "PATParser.h"
+#include "PMTParser.h"
+#include "PMTInstance.h"
+#include "PMTElementary.h"
+#include "ParseManager.h"
+
+#define TABILE_ID(buffer)   (buffer[0])
+
+ParserManager::ParserManager()
+{
+	// Add PAT parser
+	addParser(new PATParser());
+	// Add PMT parser
+	addParser(new PMTParser());
+}
+
+ParserManager::~ParserManager()
+{
+	mTableParsers.clear();
+}
+
+bool ParserManager::isPATReceived(void)
+{
+	auto pPATParser = static_cast<PATParser *>(getParser(PATParser::TABLE_ID));
+	if (!pPATParser) {
+		meddbg("PAT parser is not found!\n");
+		return false;
+	}
+
+	return pPATParser->isRecv();
+}
+
+bool ParserManager::isPMTReceived(prog_num_t progNum)
+{
+	auto pPMTParser = static_cast<PMTParser *>(getParser(PMTParser::TABLE_ID));
+	if (!pPMTParser) {
+		meddbg("PMT parser is not found!\n");
+		return false;
+	}
+
+	auto pPMTInstance = pPMTParser->getPMTInstance(progNum);
+	if (!pPMTInstance) {
+		medwdbg("PMT of program %d is not found!\n", progNum);
+		return false;
+	}
+
+	return pPMTInstance->isCompleted();
+}
+
+bool ParserManager::isPMTReceived(void)
+{
+	auto pPATParser = static_cast<PATParser *>(getParser(PATParser::TABLE_ID));
+	if (!pPATParser) {
+		meddbg("PAT parser is not found!\n");
+		return false;
+	}
+
+	if (pPATParser->isRecv()) {
+		size_t i;
+		size_t progs = pPATParser->sizeOfProgram();
+		for (i = 0; i < progs; i++) {
+			if (!isPMTReceived(pPATParser->getProgramNumber(i))) {
+				return false;
+			}
+		}
+		return true;
+	}
+
+	return false;
+}
+
+bool ParserManager::getAudioStreamInfo(prog_num_t progNum, uint8_t &streamType, ts_pid_t &pid)
+{
+	auto pPMTParser = static_cast<PMTParser *>(getParser(PMTParser::TABLE_ID));
+	auto pPMTInstance = pPMTParser->getPMTInstance(progNum);
+
+	if (pPMTInstance && pPMTInstance->isCompleted()) {
+		size_t i;
+		size_t num = pPMTInstance->numOfElementary();
+		for (i = 0; i < num; i++) {
+			auto pStream = pPMTInstance->getPMTElementary(i);
+			if (!pStream) {
+				meddbg("Run out of memory!\n");
+				return false;
+			}
+
+			switch (pStream->getStreamType()) {
+				case PMTElementary::STREAM_TYPE_AUDIO_AAC:   // fall through
+				case PMTElementary::STREAM_TYPE_AUDIO_MPEG2: // fall through
+				case PMTElementary::STREAM_TYPE_AUDIO_AC3:   // fall through
+				case PMTElementary::STREAM_TYPE_AUDIO_MPEG1: // fall through
+				case PMTElementary::STREAM_TYPE_AUDIO_HE_AAC:
+					streamType = pStream->getStreamType();
```

Comment:
we do not use streamType and pid from here, then please use direct call inside of medvdbg.

Reply:
streamType and pid is output parameters